docs: document header structure (#26624)
neovim/neovim: runtime/doc/dev_style.txt

        #pragma once
<

Headers system ~

Neovim has uses two types of headers. There are "normal" headers and "defs"
headers. Typically, each normal header will have a corresponding defs header,
e.g. fileio.h and fileio_defs.h. This distinction is done to minimize
recompilation of header files on change. Use the following rules to decide
where to put what:

- All non-symbols shall be moved to _defs.h:
  - macros, i.e. `#define`.

>     - EXCEPTION: if the macro calls a user-defined function, then it needs to

Reviewer: What is a "user-defined function"?

Author: A function defined by the user 🤪

Author: I mean non-system header

      it needs to be moved to a normal header.
  - static inline functions, but only if its function declaration has a
    REAL_FATTR_ALWAYS_INLINE attribute.
  - typedefs
  - structs

>   - enum constants

Reviewer: Note: moving an enum constant (or `#define`d constant) to a _defs cannot lead to less recompilation on change if it can only be used by the callers of a specific function (ref #26622), as users of the enum constant must have already included the header for the function, and changes to the enum constant must be associated with changes to the function.

All symbols shall be moved to regular .h:
  - regular function declarations
  - `extern` variables (including the `EXTERN` macro)

All headers are allowed to include only defs headers, system headers and
generated declarations, regardless of whether they are defs headers or regular
headers.

> c files are expected to include both _defs.h and regular headers, regular

Reviewer: ```suggestion
.c files are expected to include both _defs.h and regular headers, regular
```

        #pragma once
<

Headers system ~

> Nvim has uses two types of headers. There are "normal" headers and "defs"

Reviewer 2: ```suggestion
Nvim uses two types of headers. There are "normal" headers and "defs"
```

headers. Typically, each normal header will have a corresponding defs header,

> e.g. fileio.h and fileio_defs.h. This distinction is done to minimize

Reviewer 3: ```suggestion
e.g. `fileio.h` and `fileio_defs.h`. This distinction is done to minimize
```

> recompilation on change. The goal is to achieve the following setup:

Reviewer 2: ```suggestion
recompilation on change. The goal is to achieve the following:
```

> All headers (defs and normal) are allowed to include only defs headers, system

Reviewer 3: ```suggestion
- All headers (defs and normal) are allowed to include only defs headers, system
```

> - All headers (defs and normal) are allowed to include only defs headers, system

Reviewer 2: ```suggestion
- All headers (defs and normal) must include only defs headers, system
```

> headers and generated declarations. In other words, headers should never

Reviewer 2: ```suggestion
headers, and generated declarations. In other words, headers should never
```

Reviewer 3: "should not" or "must not" (depending on whether exceptions can be made).

> headers, and generated declarations. In other words, headers must not
> include normal headers.

Reviewer 2: ```suggestion
  headers, and generated declarations. In other words, headers must not
  include normal headers.
```

> .c files can include all headers, normal headers only if they need symbols and

Reviewer 3: ```suggestion
- .c files can include all headers but normal headers only if they need symbols and
```

> - .c files can include all headers but normal headers only if they need symbols and

Reviewer 2: I think it's a little weird to begin a sentence with punctuation. Maybe this is just me, but if others agree, I suggest this instead:

```suggestion
- Source (.c) files may include all headers, but should only include normal headers if they need symbols and
```

> not types.

Reviewer 2: ```suggestion
  not types.
```

>   not types.

Reviewer 2: This line should be wrapped

> Use the following guideline to determine where to put what:

Reviewer 3: ```suggestion
Use the following guideline to determine what to put where:
```

Symbols:
  - regular function declarations
  - `extern` variables (including the `EXTERN` macro)

Non-symbols:
  - macros, i.e. `#define`.
  - static inline functions, but only if its function declaration has a
    `REAL_FATTR_ALWAYS_INLINE` attribute.
  - typedefs
  - structs
  - enums

> • All symbols shall be moved to normal headers.

Reviewer 3: ```suggestion
- All symbols should be moved to normal headers.
```
Or "must"? (It's _really_ helpful to always stick to the common terminology "can" vs. "should" vs. "must".)

Author: I think "must" works here, thanks

> • Non-symbols used by multiple headers should be moved to defs headers. This is

Reviewer 3: ```suggestion
- Non-symbols used by multiple headers should be moved to defs headers. This is
```

> to ensure headers only include defs headers. Conversely, move non-symbols used
> by only a single header to that header.

Reviewer 3: ```suggestion
to ensure headers only include defs headers. Conversely, non-symbols used
by only a single header should be moved to that header.
```
Make it clear it's "should", not "must".

> to ensure headers only include defs headers. Conversely, non-symbols used
> by only a single header should be moved to that header.

Reviewer 2: ```suggestion
  to ensure headers only include defs headers. Conversely, non-symbols used
  by only a single header should be moved to that header.
```

> • EXCEPTION: if the macro calls a function, then it needs to be moved to a

Reviewer 3: ```suggestion
- EXCEPTION: if the macro calls a function, then it must be moved to a
```
